Public/Src/Pips/Dll/Operations/PipGraphFragmentSerializer.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.ContractsLight;
+using System.IO;
+using BuildXL.Utilities;
+
+namespace BuildXL.Pips.Operations
+{
+    /// <summary>
+    /// Class to handle serialization/deserialization of pip graph fragments
+    /// </summary>
+    public class PipGraphFragmentSerializer
+    {
+        /// <summary>
+        /// Total pips in the fragment.  Until this is read in from the file, it is 0.
+        /// </summary>
+        public int TotalPips { get; private set; }
+
+        /// <summary>
+        /// Total pips deserialized so far.
+        /// </summary>
+        public int PipsDeserialized { get; private set; }
+
+        /// <summary>
+        /// Total pips serialized so far.
+        /// </summary>
+        public int PipsSerialized { get; private set; }
+
+        /// <summary>
+        /// Name of the fragment, for printing on the console
+        /// </summary>
+        public string FragmentName { get; private set; }
+
+        /// <summary>
+        /// Deserialize a pip graph fragment and call the given handleDeserializedPip function on each pip deserialized
+        /// Returns true if successfully handled all pips.
+        /// </summary>
+        public bool Deserialize(string fragmentName, PipExecutionContext context, PipGraphFragmentContext pipFragmentContext, AbsolutePath filePath, Func<Pip, bool> handleDeserializedPip)
+        {
+            FragmentName = fragmentName;
+            PipsDeserialized = 0;
+            string fileName = filePath.ToString(context.PathTable);
+            Contract.Assert(File.Exists(fileName), "Pip graph fragment file doesn't exist");
+            using (FileStream stream = new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.Read))
+            using (RemapReader reader = new RemapReader(pipFragmentContext, stream, context))
+            {
+                while (reader.ReadBoolean())
+                {
+                    var pip = Pip.Deserialize(reader);
+                    if (handleDeserializedPip != null)
+                    {
+                        if (!handleDeserializedPip.Invoke(pip))
+                        {
+                            return false;
+                        }
+                    }
+
+                    PipsDeserialized++;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Serialize list of pips to a file
+        /// </summary>
+        public void Serialize(string fragmentName, PipExecutionContext context, AbsolutePath filePath, IEnumerable<Pip> pipsToSerialize)

```

Comment:
> Serialize [](start = 19, length = 10)

Where do we call this method?

Reply:
No where yet, this will be in PR 2 to call this.

---
In reply to: [289583923](https://github.com/microsoft/BuildXL/pull/394#discussion_r289583923) [](ancestors = 289583923)